File: Java/src/br/org/cesar/rotavital/fila/TesteFila.java
package br.org.cesar.rotavital.fila;

import br.org.cesar.rotavital.comum.TipoHemocomponente;

import static br.org.cesar.rotavital.comum.TipoHemocomponente.CRIOPRECIPITADO;
import static br.org.cesar.rotavital.comum.TipoHemocomponente.HEMACIAS;
import static br.org.cesar.rotavital.comum.TipoHemocomponente.PLAQUETAS;
import static br.org.cesar.rotavital.comum.TipoHemocomponente.PLASMA;

/**
 * Casos de teste da fila, equivalentes a teste_fila.c, com a mesma saída [PASSOU]/[FALHOU].
 * Sem dependências (sem JUnit) para rodar só com javac/java.
 */
public class TesteFila {

    private static int casosExecutados = 0;
    private static int casosFalhos = 0;

    private interface Caso {
        void executar();
    }

    private static void verificar(boolean condicao, String descricao) {
        if (!condicao) {
            throw new AssertionError(descricao);
        }
    }

    private static void rodar(Caso caso, String descricao) {
        casosExecutados++;
        try {
            caso.executar();
            System.out.println("  [PASSOU] " + descricao);
        } catch (AssertionError | RuntimeException e) {
            casosFalhos++;
            System.out.println("  [FALHOU] " + descricao + "\n           -> " + e);
        }
    }

    private static Requisicao novaRequisicao(String hospital, String tipo,
                                             TipoHemocomponente componente, int quantidade) {
        return new Requisicao(hospital, "Dra. Plantonista", tipo, componente, quantidade, "2026-09-23");
    }

    private static FilaRequisicoes filaComTres() {
        FilaRequisicoes fila = new FilaRequisicoes();
        fila.enfileirar(novaRequisicao("IMIP", "O-", HEMACIAS, 2));
        fila.enfileirar(novaRequisicao("Hospital da Restauracao", "A+", PLASMA, 1));
        fila.enfileirar(novaRequisicao("IMIP", "B-", PLAQUETAS, 3));
        return fila;
    }

    private static boolean rejeita(FilaRequisicoes fila, Requisicao requisicao) {
        try {
            fila.enfileirar(requisicao);
            return false;
        } catch (IllegalArgumentException e) {
            return true;
        }
    }

    /* ---------- Criar ---------- */

    private static void criarFilaVazia() {
        FilaRequisicoes fila = new FilaRequisicoes();
        verificar(fila.vazia(), "fila nova vazia");
        verificar(fila.tamanho() == 0, "tamanho 0");
        verificar(fila.frente() == null, "sem frente");
    }

    /* ---------- Inserir (enqueue) ---------- */

    private static void enfileirarEmFilaVazia() {
        FilaRequisicoes fila = new FilaRequisicoes();
        int id = fila.enfileirar(novaRequisicao("IMIP", "O-", HEMACIAS, 2));

        verificar(id == 1, "primeiro id e 1");
        verificar(fila.tamanho() == 1, "tamanho 1");
        verificar(fila.frente().getId() == 1, "unico item e a frente");
    }

    private static void enfileirarVaiParaOFim() {
        FilaRequisicoes fila = new FilaRequisicoes();
        fila.enfileirar(novaRequisicao("IMIP", "O-", HEMACIAS, 2));
        int id = fila.enfileirar(novaRequisicao("IMIP", "A+", PLASMA, 1));

        verificar(id == 2, "segundo id e 2");
        verificar(fila.frente().getId() == 1, "frente nao muda");
        verificar(fila.posicao(2) == 2, "novo item fica no fim");
    }

    private static void enfileirarGuardaCopia() {
        FilaRequisicoes fila = new FilaRequisicoes();
        Requisicao original = novaRequisicao("IMIP", "O-", HEMACIAS, 2);
        fila.enfileirar(original);
        original.setHospital("Alterado depois");

        verificar(fila.frente().getHospital().equals("IMIP"), "fila nao e afetada pelo objeto original");
        fila.frente().setQuantidade(99);
        verificar(fila.frente().getQuantidade() == 2, "frente() devolve copia");
    }

    private static void enfileirarDadosInvalidosERejeitado() {
        FilaRequisicoes fila = new FilaRequisicoes();
        Requisicao dataRuim = novaRequisicao("IMIP", "O+", HEMACIAS, 1);
        Requisicao semResponsavel = novaRequisicao("IMIP", "O+", HEMACIAS, 1);
        Requisicao semComponente = novaRequisicao("IMIP", "O+", null, 1);
        dataRuim.setData("2026-13-01");
        semResponsavel.setResponsavel("");

        verificar(rejeita(fila, novaRequisicao("", "O+", HEMACIAS, 1)), "sem hospital");
        verificar(rejeita(fila, novaRequisicao("IMIP", "X", HEMACIAS, 1)), "tipo invalido");
        verificar(rejeita(fila, novaRequisicao("IMIP", "O+", HEMACIAS, -1)), "quantidade negativa");
        verificar(rejeita(fila, dataRuim), "data invalida");
        verificar(rejeita(fila, semResponsavel), "sem responsavel");
        verificar(rejeita(fila, semComponente), "sem componente");
        verificar(rejeita(fila, null), "requisicao nula");
        verificar(fila.vazia(), "nada foi inserido");
        verificar(fila.enfileirar(novaRequisicao("IMIP", "O+", HEMACIAS, 1)) == 1,
                "rejeicoes nao consomem id");
    }

    /* ---------- Remover (dequeue) ---------- */

    private static void desenfileirarEmOrdemFifo() {
        FilaRequisicoes fila = filaComTres();

        Requisicao saida = fila.desenfileirar();
        verificar(saida.getId() == 1, "sai o id 1");
        verificar(saida.getHospital().equals("IMIP"), "dados preservados");
        verificar(fila.desenfileirar().getId() == 2, "sai o id 2");
        verificar(fila.desenfileirar().getId() == 3, "sai o id 3");
    }

    private static void desenfileirarUltimoEsvaziaFila() {
        FilaRequisicoes fila = new FilaRequisicoes();
        fila.enfileirar(novaRequisicao("IMIP", "O-", HEMACIAS, 2));
        fila.desenfileirar();

        verificar(fila.vazia(), "fila vazia");
        verificar(fila.tamanho() == 0, "tamanho 0");
        verificar(fila.frente() == null, "sem frente");
    }

    private static void desenfileirarFilaVaziaLancaExcecao() {
        FilaRequisicoes fila = new FilaRequisicoes();
        try {
            fila.desenfileirar();
            verificar(false, "deveria lancar FilaVaziaException");
        } catch (FilaVaziaException e) {
            verificar(e.getMessage().equals("Fila vazia"), "mensagem da excecao");
        }
    }

    private static void reutilizarFilaAposEsvaziar() {
        FilaRequisicoes fila = new FilaRequisicoes();
        Requisicao a = novaRequisicao("IMIP", "AB-", CRIOPRECIPITADO, 1);
        fila.enfileirar(a);
        fila.desenfileirar();

        verificar(fila.enfileirar(a) == 2, "ids nao sao reaproveitados");
        verificar(fila.tamanho() == 1, "tamanho 1");
        verificar(fila.frente().getId() == 2, "novo item e a frente");
    }

    /* ---------- Consultar ---------- */

    private static void consultarFrenteSemRemover() {
        FilaRequisicoes fila = filaComTres();
        Requisicao r = fila.frente();

        verificar(r != null && r.getId() == 1, "frente e o id 1");
        verificar(fila.tamanho() == 3, "frente nao remove");
    }

    private static void consultarPorId() {
        FilaRequisicoes fila = filaComTres();
        Requisicao r = fila.buscar(3);

        verificar(r != null, "encontra id 3");
        verificar(r.getTipoSanguineo().equals("B-"), "dados do id 3");
        verificar(fila.buscar(99) == null, "id inexistente");
    }

    private static void consultarPosicaoNaFila() {
        FilaRequisicoes fila = filaComTres();

        verificar(fila.posicao(1) == 1, "id 1 na posicao 1");
        verificar(fila.posicao(3) == 3, "id 3 na posicao 3");
        verificar(fila.posicao(99) == 0, "inexistente retorna 0");
        fila.desenfileirar();
        verificar(fila.posicao(3) == 2, "todos avancam uma posicao");
    }

    private static void consultarPendenciasPorHospital() {
        FilaRequisicoes fila = filaComTres();

        verificar(fila.contarPorHospital("IMIP") == 2, "IMIP tem 2");
        verificar(fila.contarPorHospital("Hospital Inexistente") == 0, "inexistente tem 0");
        fila.desenfileirar();
        verificar(fila.contarPorHospital("IMIP") == 1, "IMIP fica com 1");
    }

    private static void listarEmOrdemDeAtendimento() {
        FilaRequisicoes fila = filaComTres();
        java.util.List<Requisicao> lista = fila.listar();

        verificar(lista.size() == 3, "lista com 3");
        verificar(lista.get(0).getId() == 1 && lista.get(2).getId() == 3, "ordem FIFO");
    }

    public static void main(String[] args) {
        System.out.println("\n=== Testes: Requisicoes hospitalares (fila) - Java ===");

        System.out.println("\n Criar");
        rodar(TesteFila::criarFilaVazia, "Criar fila vazia");

        System.out.println("\n Inserir (enqueue)");
        rodar(TesteFila::enfileirarEmFilaVazia, "Enfileirar em fila vazia");
        rodar(TesteFila::enfileirarVaiParaOFim, "Enfileirar vai para o fim");
        rodar(TesteFila::enfileirarGuardaCopia, "Enfileirar guarda uma copia da requisicao");
        rodar(TesteFila::enfileirarDadosInvalidosERejeitado, "Enfileirar dados invalidos e rejeitado");

        System.out.println("\n Remover (dequeue)");
        rodar(TesteFila::desenfileirarEmOrdemFifo, "Desenfileirar em ordem de chegada (FIFO)");
        rodar(TesteFila::desenfileirarUltimoEsvaziaFila, "Desenfileirar o ultimo esvazia a fila");
        rodar(TesteFila::desenfileirarFilaVaziaLancaExcecao, "Desenfileirar fila vazia lanca excecao");
        rodar(TesteFila::reutilizarFilaAposEsvaziar, "Reutilizar a fila depois de esvaziar");

        System.out.println("\n Consultar");
        rodar(TesteFila::consultarFrenteSemRemover, "Consultar a frente sem remover");
        rodar(TesteFila::consultarPorId, "Consultar por id");
        rodar(TesteFila::consultarPosicaoNaFila, "Consultar posicao na fila");
        rodar(TesteFila::consultarPendenciasPorHospital, "Consultar pendencias por hospital");
        rodar(TesteFila::listarEmOrdemDeAtendimento, "Listar em ordem de atendimento");

        System.out.printf("%nResultado: %d/%d casos passaram%n", casosExecutados - casosFalhos, casosExecutados);
        System.exit(casosFalhos == 0 ? 0 : 1);
    }
}

File: C/teste_fila.c
#include "fila.h"
#include "teste.h"

#include <string.h>

static Requisicao nova_requisicao(const char *hospital, const char *tipo,
                                  TipoHemocomponente componente, int quantidade) {
    Requisicao requisicao;
    memset(&requisicao, 0, sizeof(requisicao));
    strncpy(requisicao.hospital, hospital, TAM_HOSPITAL - 1);
    strncpy(requisicao.responsavel, "Dra. Plantonista", TAM_RESPONSAVEL - 1);
    strncpy(requisicao.tipoSanguineo, tipo, TAM_TIPO_SANGUINEO - 1);
    requisicao.componente = componente;
    requisicao.quantidade = quantidade;
    strcpy(requisicao.data, "2026-09-23");
    return requisicao;
}

static FilaRequisicoes *fila_com_tres(void) {
    FilaRequisicoes *fila = fila_criar();
    Requisicao a = nova_requisicao("IMIP", "O-", HEMACIAS, 2);
    Requisicao b = nova_requisicao("Hospital da Restauracao", "A+", PLASMA, 1);
    Requisicao c = nova_requisicao("IMIP", "B-", PLAQUETAS, 3);
    fila_enfileirar(fila, &a, NULL);
    fila_enfileirar(fila, &b, NULL);
    fila_enfileirar(fila, &c, NULL);
    return fila;
}

/* ---------- Criar / destruir ---------- */

static void criar_fila_vazia(void) {
    FilaRequisicoes *fila = fila_criar();
    VERIFICAR(fila != NULL);
    VERIFICAR(fila->inicio == NULL);
    VERIFICAR(fila->fim == NULL);
    VERIFICAR(fila->tamanho == 0);
    VERIFICAR(fila_vazia(fila));
    fila_destruir(&fila);
}

static void destruir_zera_ponteiro(void) {
    FilaRequisicoes *fila = fila_com_tres();
    fila_destruir(&fila);
    VERIFICAR(fila == NULL);
    fila_destruir(&fila); /* destruir duas vezes não deve quebrar */
}

/* ---------- Inserir (enqueue) ---------- */

static void enfileirar_em_fila_vazia(void) {
    FilaRequisicoes *fila = fila_criar();
    Requisicao a = nova_requisicao("IMIP", "O-", HEMACIAS, 2);
    int id = 0;

    VERIFICAR(fila_enfileirar(fila, &a, &id) == FILA_OK);
    VERIFICAR(id == 1);
    VERIFICAR(fila->inicio != NULL);
    VERIFICAR(fila->inicio == fila->fim); /* um único nó: início e fim coincidem */
    VERIFICAR(fila->fim->proximo == NULL);
    VERIFICAR(fila->tamanho == 1);
    fila_destruir(&fila);
}

static void enfileirar_avanca_so_o_fim(void) {
    FilaRequisicoes *fila = fila_criar();
    Requisicao a = nova_requisicao("IMIP", "O-", HEMACIAS, 2);
    Requisicao b = nova_requisicao("IMIP", "A+", PLASMA, 1);
    NoRequisicao *inicioOriginal;
    int id = 0;

    fila_enfileirar(fila, &a, NULL);
    inicioOriginal = fila->inicio;
    VERIFICAR(fila_enfileirar(fila, &b, &id) == FILA_OK);
    VERIFICAR(id == 2);
    VERIFICAR(fila->inicio == inicioOriginal); /* início não muda */
    VERIFICAR(fila->fim->requisicao.id == 2);
    VERIFICAR(fila->inicio->proximo == fila->fim);
    VERIFICAR(fila->fim->proximo == NULL);
    fila_destruir(&fila);
}

static void enfileirar_dados_invalidos_e_rejeitado(void) {
    FilaRequisicoes *fila = fila_criar();
    Requisicao semHospital = nova_requisicao("", "O+", HEMACIAS, 1);
    Requisicao tipoRuim = nova_requisicao("IMIP", "X", HEMACIAS, 1);
    Requisicao qtdNegativa = nova_requisicao("IMIP", "O+", HEMACIAS, -1);
    Requisicao dataRuim = nova_requisicao("IMIP", "O+", HEMACIAS, 1);
    Requisicao semResponsavel = nova_requisicao("IMIP", "O+", HEMACIAS, 1);
    strcpy(dataRuim.data, "2026-13-01");
    semResponsavel.responsavel[0] = '\0';

    VERIFICAR(fila_enfileirar(fila, &semHospital, NULL) == FILA_PARAMETRO_INVALIDO);
    VERIFICAR(fila_enfileirar(fila, &tipoRuim, NULL) == FILA_PARAMETRO_INVALIDO);
    VERIFICAR(fila_enfileirar(fila, &qtdNegativa, NULL) == FILA_PARAMETRO_INVALIDO);
    VERIFICAR(fila_enfileirar(fila, &dataRuim, NULL) == FILA_PARAMETRO_INVALIDO);
    VERIFICAR(fila_enfileirar(fila, &semResponsavel, NULL) == FILA_PARAMETRO_INVALIDO);
    VERIFICAR(fila_enfileirar(fila, NULL, NULL) == FILA_PARAMETRO_INVALIDO);
    VERIFICAR(fila_enfileirar(NULL, &tipoRuim, NULL) == FILA_PARAMETRO_INVALIDO);
    VERIFICAR(fila_vazia(fila));
    VERIFICAR(fila->proximoId == 1); /* rejeições não consomem id */
    fila_destruir(&fila);
}

/* ---------- Remover (dequeue) ---------- */

static void desenfileirar_em_ordem_fifo(void) {
    FilaRequisicoes *fila = fila_com_tres();
    Requisicao saida;

    VERIFICAR(fila_desenfileirar(fila, &saida) == FILA_OK);
    VERIFICAR(saida.id == 1);
    VERIFICAR(strcmp(saida.hospital, "IMIP") == 0);
    VERIFICAR(fila_desenfileirar(fila, &saida) == FILA_OK);
    VERIFICAR(saida.id == 2);
    VERIFICAR(fila_desenfileirar(fila, &saida) == FILA_OK);
    VERIFICAR(saida.id == 3);
    fila_destruir(&fila);
}

static void desenfileirar_ultimo_zera_inicio_e_fim(void) {
    FilaRequisicoes *fila = fila_criar();
    Requisicao a = nova_requisicao("IMIP", "O-", HEMACIAS, 2);
    fila_enfileirar(fila, &a, NULL);

    VERIFICAR(fila_desenfileirar(fila, NULL) == FILA_OK);
    VERIFICAR(fila->inicio == NULL);
    VERIFICAR(fila->fim == NULL); /* fim não pode apontar para o nó liberado */
    VERIFICAR(fila->tamanho == 0);
    fila_destruir(&fila);
}

static void desenfileirar_fila_vazia_retorna_erro(void) {
    FilaRequisicoes *fila = fila_criar();
    Requisicao saida;
    VERIFICAR(fila_desenfileirar(fila, &saida) == FILA_VAZIA);
    VERIFICAR(fila_desenfileirar(NULL, &saida) == FILA_PARAMETRO_INVALIDO);
    fila_destruir(&fila);
}

static void reutilizar_fila_apos_esvaziar(void) {
    FilaRequisicoes *fila = fila_criar();
    Requisicao a = nova_requisicao("IMIP", "AB-", CRIOPRECIPITADO, 1);
    int id = 0;

    fila_enfileirar(fila, &a, NULL);
    fila_desenfileirar(fila, NULL);
    VERIFICAR(fila_enfileirar(fila, &a, &id) == FILA_OK);
    VERIFICAR(id == 2); /* ids não são reaproveitados */
    VERIFICAR(fila->inicio == fila->fim);
    VERIFICAR(fila->inicio != NULL);
    fila_destruir(&fila);
}

/* ---------- Consultar ---------- */

static void consultar_frente_sem_remover(void) {
    FilaRequisicoes *fila = fila_com_tres();
    FilaRequisicoes *vazia = fila_criar();
    const Requisicao *r = fila_frente(fila);

    VERIFICAR(r != NULL);
    VERIFICAR(r->id == 1);
    VERIFICAR(fila->tamanho == 3);
    VERIFICAR(fila_frente(vazia) == NULL);
    fila_destruir(&fila);
    fila_destruir(&vazia);
}

static void consultar_por_id(void) {
    FilaRequisicoes *fila = fila_com_tres();
    const Requisicao *r = fila_buscar(fila, 3);

    VERIFICAR(r != NULL);
    VERIFICAR(strcmp(r->tipoSanguineo, "B-") == 0);
    VERIFICAR(fila_buscar(fila, 99) == NULL);
    fila_destruir(&fila);
}

static void consultar_posicao_na_fila(void) {
    FilaRequisicoes *fila = fila_com_tres();

    VERIFICAR(fila_posicao(fila, 1) == 1);
    VERIFICAR(fila_posicao(fila, 3) == 3);
    VERIFICAR(fila_posicao(fila, 99) == 0);
    fila_desenfileirar(fila, NULL);
    VERIFICAR(fila_posicao(fila, 3) == 2); /* todos avançam uma posição */
    fila_destruir(&fila);
}

static void consultar_pendencias_por_hospital(void) {
    FilaRequisicoes *fila = fila_com_tres();

    VERIFICAR(fila_contar_por_hospital(fila, "IMIP") == 2);
    VERIFICAR(fila_contar_por_hospital(fila, "Hospital Inexistente") == 0);
    fila_desenfileirar(fila, NULL);
    VERIFICAR(fila_contar_por_hospital(fila, "IMIP") == 1);
    fila_destruir(&fila);
}

int main(void) {
    teste_titulo("Testes: Requisicoes hospitalares (fila)");

    teste_secao("Criar / destruir");
    RODAR(criar_fila_vazia, "Criar fila vazia");
    RODAR(destruir_zera_ponteiro, "Destruir libera os nos e zera o ponteiro");

    teste_secao("Inserir (enqueue)");
    RODAR(enfileirar_em_fila_vazia, "Enfileirar em fila vazia (inicio == fim)");
    RODAR(enfileirar_avanca_so_o_fim, "Enfileirar avanca so o fim");
    RODAR(enfileirar_dados_invalidos_e_rejeitado, "Enfileirar dados invalidos e rejeitado");

    teste_secao("Remover (dequeue)");
    RODAR(desenfileirar_em_ordem_fifo, "Desenfileirar em ordem de chegada (FIFO)");
    RODAR(desenfileirar_ultimo_zera_inicio_e_fim, "Desenfileirar o ultimo zera inicio e fim");
    RODAR(desenfileirar_fila_vazia_retorna_erro, "Desenfileirar fila vazia retorna erro");
    RODAR(reutilizar_fila_apos_esvaziar, "Reutilizar a fila depois de esvaziar");

    teste_secao("Consultar");
    RODAR(consultar_frente_sem_remover, "Consultar a frente sem remover");
    RODAR(consultar_por_id, "Consultar por id");
    RODAR(consultar_posicao_na_fila, "Consultar posicao na fila");
    RODAR(consultar_pendencias_por_hospital, "Consultar pendencias por hospital");

    return teste_resumo();
}
